pkg/cloud-init/cloud-init.go

```diff
@@ -202,12 +213,52 @@ func GenerateLocalData(vmiName string, hostname string, namespace string, source
 	} else {
 		return fmt.Errorf("userDataBase64 or userData is required for no-cloud data source")
 	}
+
+	var preNetworkData []byte
+	if source.NetworkData != "" {
+		preNetworkData = []byte(source.NetworkData)
+	} else if source.NetworkDataBase64 != "" {
+		preNetworkData, err = base64.StdEncoding.DecodeString(source.NetworkDataBase64)
+		if err != nil {
+			return err
+		}
+	}
+
+	var networkData []byte
+	var resolvData []byte
+	if len(preNetworkData) > 0 {
+		parsedNetworkData := bytes.Split(preNetworkData, []byte(v1.CloudInitDelimiter))
+		networkData = parsedNetworkData[0]
+		if len(parsedNetworkData) > 1 {
+			resolvData = parsedNetworkData[1]
+		}
+	}
+
 	metaData := []byte(fmt.Sprintf("{ \"instance-id\": \"%s.%s\", \"local-hostname\": \"%s\" }\n", vmiName, namespace, hostname))
 
 	diskutils.RemoveFile(userFile)
 	diskutils.RemoveFile(metaFile)
+	diskutils.RemoveFile(networkFile)
 	diskutils.RemoveFile(isoStaging)
 
+	// If we have network data and userData is of type #cloud-config
+	// then append resolv configuration to userData
+	if len(networkData) > 0 && len(resolvData) > 0 {
```

Comment:
why are we altering the userdata here? Is this functionality defined in the cloud-init spec? I don't see any information about this ###CLOUDINITDELIMITER### either

Reply:
This was discussed here: https://github.com/kubevirt/kubevirt/pull/1806#discussion_r241986071

Reply:
Actually, various discussions on a few questions. I'll try to summarize the current state of things here:

- This code (as a whole) is only run when:
  - The feature gated SR-IOV functionality is enabled
  - An SR-IOV interface is present
  - networkData is not already defined
  - Only targets configuring SR-IOV interfaces via cloud-init network-config
- Resolv data:
  - Is the same data presented by the dhcp service
  - Will likely be overwritten by the VM's OS dhcp client scripts if a supported dhcp interface is present.
  - Will only be appended to userData if
    - userData is of type #cloud-config
    - `manage_resolv_conf:` does not appear anywhere in userData (putting in a comment or setting this to true or false in userData will disable adding resolv data)

We couldn't use the nameserver type in network-config because the cloud-init code appends instead of replacing the resolv configuration. Many upstream cloud images have a default configuration in place which causes failures or resolution delays.

We are also only configuring a single SR-IOV interface on our VMIs so we can't depend on setting resolv configurations through a dhcp supported interface type. (see #1807)

I started out adding another type to the API for this discovered data to be passed around with, but it seemed wrong in multiple ways. 

Let me know if you have any more questions.

Reply:
Nothing outside of the auto discovery code needs to know about ###CLOUDINITDELIMITER###.  I marshal the the network-config data and resolv data as separate objects and munge them into the byte array returned from `CloudInitDiscoverNetworkData` in `pkg/virt-launcher/virtwrap/network/cloud-init-network.go` to be split apart later.

I could make another struct to encompass both of these under but then I'd need to know about the struct in `pkg/cloud-init/cloud-init.go` to unmarshal it and write out the separate parts, or make a struct that both packages know about and pass it as an object. Although trying to include various packages in other packages within kubevirt/pkgs causes dependency loops as I believe I went down this path at the start for something else.

Reply:
why can't the user append their own resolv conf to the user data rather than us attempt to inject it from the network data?

I don't feel good about us altering any user data. We should be treating the userdata as opaque values.

Reply:
If we are auto discovering for SR-IOV we should use what was configured in the pod. If we wanted users to manually configure resolv why do we supply it with dhcp? . It is easy enough to disable the behavior with a comment or a valid manage_resolv_conf in the userData or I can add an explicit flag to the network section of the spec to turn it on/off along with a user definable global setting.

Reply:
> If we are auto discovering for SR-IOV we should use what was configured in the pod. If we wanted users to manually configure resolv why do we supply it with dhcp?

I see. So the resolv.conf data being injected into the userdata is something dynamically detected within the pods environment. I missed that the first time around.

Functionally, I get what you're trying to achieve now. It's still unclear why you're wanting to append resolve data to the userdata though.

This is the same data we're supplying via DHCP. Is this just for a scenario where a dhcp client isn't available? I'm unclear on the intent here.

Reply:
We are doing VMI with only a single SR-IOV interface so there is no bridged interface on which to dhcp.

Reply:
> We are doing VMI with only a single SR-IOV interface so there is no bridged interface on which to dhcp.

ah yes, that was the missing piece. I get why you're wanting this now.  I'm not sure we'll agree to support inspecting and altering cloud-init userdata automatically like you're proposing here though. 

Here's what I suggest. Lets get merged what we do agree on, and that's the support of the network config userdata without the CLOUDINITDELIMITER resolv.conf injection stuff.

For that second part resolv.conf part, I suggest bringing this topic up on the mailing list for further discussion. It's possible we should consider a webhook for dynamically altering cloud-init at runtime similar to the way we allow people to modify the libvirt domain xml. That would allow you to inject dynamic data into cloud-init from within the pod's environment at the point the VM is about to start.

Reply:
How about a switch to turn it on/off, if I can't get this upstreamed I'll be maintaining my own branch anyway. How many are actually using SR-IOV at this point anyway? It is still defaulted to disabled via feature gate.

Reply:
It should also be noted that I'm not altering the source of userData, just the file that is written out and packaged in the ISO.

It only happens when the user:

1. Has explicitly enable the SRIOV feature gate
2. Has explicitly enable the NetconfAutoResolv feature gate (I just added this).
3. Has an actual working SRIOV setup (at least well enough to acquire a VF and start the launcher pod)
4.  Hasn't provided networkData of their own (even if only to disable SR-IOV network auto discovery, EG '#')
5. Has provided userData and it is of type #cloud-config
5. Hasn't included manage_resolv_conf anywhere in userData

If you'd like I can also skip resolv data if there are any dhcp capable interface types seen during the discovery.

I've strived to make sure none of this code would affect any pre SR-IOV support configurations and given options to disable it for any early adopters that might have running/working setups now. Not being able to configure resolv data on SR-IOV only VMIs via cloud-init is a non starter for us.

Reply:
Ditched feature gate and implemented skip resolv data if there are any dhcp capable interface types present.

Reply:
> It should also be noted that I'm not altering the source of userData, just the file that is written out and packaged in the ISO.

yes, I get that. This still crosses the boundary of treating cloud init userdata as an opaque value. That's the crux of the issue. 

I'm not suggesting we don't support your use case. We need to look at alternative ways to achieve what you're wanting to do here. As I pointed out previously, a cloud-init hook would be a reasonable option that we can support.

Reply:
Spoke with @davidvossel offline for clarification. I'm going to put this PR on hold for now while I investigate adding a hook to pkg/cloud-init. I'll likely split out the non SR-IOV portions of the network-config api changes into a new/clean PR/branch.